server-common/src/main/java/org/apache/kafka/server/config/ServerTopicConfigSynonyms.java

```diff
@@ -0,0 +1,140 @@
+        sameNameWithLogPrefix(TopicConfig.PREALLOCATE_CONFIG),
+        sameNameWithLogPrefix(TopicConfig.MESSAGE_FORMAT_VERSION_CONFIG),
+        sameNameWithLogPrefix(TopicConfig.MESSAGE_TIMESTAMP_TYPE_CONFIG),
+        sameNameWithLogPrefix(TopicConfig.MESSAGE_TIMESTAMP_DIFFERENCE_MAX_MS_CONFIG),
+        sameNameWithLogPrefix(TopicConfig.MESSAGE_DOWNCONVERSION_ENABLE_CONFIG)
+    ));
+
+    /**
+     * Map topic config to the server config with the highest priority. Some of these have additional
+     * synonyms that can be obtained using [[kafka.server.DynamicBrokerConfig#brokerConfigSynonyms]]
+     * or using [[AllTopicConfigSynonyms]]
+     */
+    public static final Map<String, String> TOPIC_CONFIG_SYNONYMS = Collections.unmodifiableMap(
+        ALL_TOPIC_CONFIG_SYNONYMS.entrySet()
+        .stream()
+        .collect(Collectors.toMap(Entry::getKey, e -> e.getValue().get(0).name())));
+
+    /**
+     * Return the server config with the highest priority for `topicConfigName` if it exists. Otherwise,
+     * throw NoSuchElementException.
+     */
+    public static String serverSynonym(String topicConfigName) {
+        String serverSynonym = TOPIC_CONFIG_SYNONYMS.get(topicConfigName);
+        if (serverSynonym == null)
+            throw new NoSuchElementException("No server synonym found for " + topicConfigName);
+        return serverSynonym;
+    }
+
+    private static Entry<String, List<ConfigSynonym>> sameName(String configName) {
+        return Utils.mkEntry(configName, asList(new ConfigSynonym(configName)));
+    }
+
+    private static Entry<String, List<ConfigSynonym>> sameNameWithLogPrefix(String configName) {
+        return Utils.mkEntry(configName, asList(new ConfigSynonym(LOG_PREFIX + configName)));
+    }
+
+    private static Entry<String, List<ConfigSynonym>> sameNameWithLogCleanerPrefix(String configName) {
+        return Utils.mkEntry(configName, asList(new ConfigSynonym(LOG_CLEANER_PREFIX + configName)));
+    }
+
+    private static Entry<String, List<ConfigSynonym>> singleWithLogPrefix(String topicConfigName, String brokerConfigName) {
+        return Utils.mkEntry(topicConfigName, asList(new ConfigSynonym(LOG_PREFIX + brokerConfigName)));
+    }
+
+    private static Entry<String, List<ConfigSynonym>> singleWithLogCleanerPrefix(String topicConfigName, String brokerConfigName) {
+        return Utils.mkEntry(topicConfigName, asList(new ConfigSynonym(LOG_CLEANER_PREFIX + brokerConfigName)));
+    }
+
+    private static Entry<String, List<ConfigSynonym>> listWithLogPrefix(String topicConfigName, ConfigSynonym... synonyms) {
+        List<ConfigSynonym> synonymsWithPrefix = Arrays.stream(synonyms)
+            .map(s -> new ConfigSynonym(LOG_PREFIX + s.name(), s.converter()))
+            .collect(Collectors.toList());
+        return Utils.mkEntry(topicConfigName, synonymsWithPrefix);
+    }
+
+    private static Entry<String, List<ConfigSynonym>> single(String topicConfigName, String brokerConfigName) {
```

Comment:
Good question. The naming scheme was trying to cover 3 different cases, each with prefix/no prefix variants:

1. Single mapping between topic and broker config name: `single` (this is like `Collections.singleton` in a way)
2. Mapping of topic to multiple broker synonyms: `list`
3. A mapping where the topic and broker config have the same name: `sameName` (this could also have been called `identity` now that I think of it).

Are you suggesting that all `single*` methods become `differentName*`? What about `list*`, does that become `differentNameList*`?

Reply:
Thanks for the explanation. Got it. I am not sure that I could make the naming better. We can just keep it as it is.